core/src/main/java/org/eclipse/hono/util/NotificationDeviceReadyToDeliver.java

```diff
@@ -0,0 +1,61 @@
+/**
+ * Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 1.0 which is available at
+ * https://www.eclipse.org/legal/epl-v10.html
+ *
+ * SPDX-License-Identifier: EPL-1.0
+ */
+
+package org.eclipse.hono.util;
+
+import java.time.Instant;
+
+/**
+ * Contains all information about a device that is indicating being ready to receive an upstream message.
+ */
+public final class NotificationDeviceReadyToDeliver {
```

Comment:
what is this good for?

Reply:
This defines the values passed to the callback method in the business application for convenience in one class.
The idea is to have everything contained to enable the application to send a command upstream.

Like
```
private void handleCommandReadinessNotification(final NotificationDeviceReadyToDeliver notificationDeviceReadyToDeliver) {
  // do something here, e.g. send a command
}```

Reply:
since there is no such thing as a *CommandReadinessNotification* I do not see how this is supposed to work ...

Reply:
This is generated by the demuxer that is evaluating the `piggy-backed` message and then invokes
the callback with this object (which is constructed from the message).

```
Optional.ofNullable(notificationReadyToDeliverConsumer).map(notificationConsumer -> {
                getReadyForCommandsNotification(tenantId, msg).ifPresent(
                        notificationObject -> notificationConsumer.accept(notificationObject));
. . .
```

This is how such an object is put into the callback implemented in the example already, without having a specific `CommandReadinessNotification`.
The application itself does not have to demux/switch after receiving a message by calling helper methods, but instead only needs to provide the callback method to the demuxer.

By this Hono can internally determine  when it thinks a device is ready to receive a command and notify the application  - no business coding involved.

Reply:
Ok, I think I get it now. The demuxer is used to detect all messages that contain a ttd property and if they do, creates an instance of this class and forwards it to the notificationConsumer while forwarding the original message to the downstream consumer, right?

However, we definitely need to find a new (better) name for this class. There is nothing being delivered here, is there?  What about `TimeTilDisconnectNotification`?

Reply:
> Ok, I think I get it now. The demuxer is used to detect all messages that contain a ttd property and if they do, creates an instance of this class and forwards it to the notificationConsumer while forwarding the original message to the downstream consumer, right?

Exactly!

> However, we definitely need to find a new (better) name for this class.

:+1: 

> There is nothing being delivered here, is there? What about TimeTilDisconnectNotification?

Thanks for the proposal, I wasn't too happy with the name either...
`TimeUntilDisconnectNotification` currently looks best for me - reducing the 'til' only to the protocol header names. WDYT?